docs/design/sql/02-operator-interface.md

```diff
@@ -0,0 +1,163 @@
+---- GroupBy [a]
+---- Having [SUM(b) > 50]
+```
+*Snippet 3: Relational Tree*
+```
+-- Filter [SUM(b) > 50]
+---- Aggregate [a -> SUM(b)]
+------ Project [a, b]
+-------- Scan [table]
+```
+
+## Volcano Model
+
+Volcano Model defines the common data exchange interface between operators in the relational tree. This allows
+for extensibility, as new operators could be implemented with minimal changes to the engine.
+
+In the original paper the interface consists of three operations:
+
+*Snippet 4: Volcano Interface*
+```java
+interface Operator {
+    void open();  // Initialize the operator
+    Row next();   // Get the next row
+    void close(); // Close the operator and release all resources
+}
+```
+
+## Volcano Model in Hazelcast Mustang
+
+The original Volcano Model has two drawbacks:
+1. Operators exchange one row at a time, which leads to performance overhead
+2. Call to the `next()` is blocking, which is not optimal for the distributed environment, where
+operators often wait for remote data or free space in the send buffer.
+
+To achieve high performance, we introduce several changes to the original Volcano Model: batching and
+non-blocking execution.
+
+### Row and RowBatch
+We define the `RowBatch` interface which a collection of rows (tuples).
+
+*Snippet 5: RowBatch interface*
+```java
+interface RowBatch {
+    Row getRow(int index); // Get the row by index
+    int getRowCount();     // Get the number of rows 
+} 
+```
+
+Then we define the `Row` interface, which provides access to values by index. The `Row` itself is considered
+as a special case of `RowBatch` with one row. This allows saving on allocations in some parts of the engine.
+
+*Snippet 6: Row interface*
+```java
+interface Row extends RowBatch {
+    Object get(int index); // Get the value by index
+    int getColumnCount();  // Get the number of values in the row 
+    
+    default int getRowCount() {
+        return 1;
+    }
+    
+    default Row getRow(int index) {
+        return this;
+    }
+}
+```
+
+### Operator
+The operator is defined by `Exec` interface:
+1. Operators exchange `RowBatch` instead of `Row`
+1. The blocking `next()` method is replaced with the non-blocking `advance()` method, which returns the iteration
+result instead of the row batch
+1. The `RowBatch` could be accessed through a separate method
+1. The `open()` method is renamed to `setup()`. Special query context is passed to it as an argument
+1. There is no separate `close()` method because the engine doesn't need explicit per-operator cleanup at the
+moment. This may change in future, in this case the current document should be updated accordingly
+
+*Snippet 7: Executable operator interface*
+```java
+interface Exec {
+    void setup(QueryContext context); // Initialize the operator
+    IterationResult advance();        // Advance the operator if possible; never blocks
+    RowBatch currentBatch();          // Get the batch returned by the previous advance() call 
```

Comment:
Just as an idea, these two methods can be combined into a single one: `RowBatch  advance()` returning special signaling `RowBatch` instances named the same and having the same semantics as `IterationResult` members.

Reply:
This will require adding some additional state to the `RowBatch`, which is not concerned with the batch itself, but with the operator which produced it. I would avoid mixing the concepts.

Reply:
Not really:

Reply:
> This will require adding some additional state to the RowBatch

Not really:

```java
final class SignalingRowBatch implements RowBatch {

    public static final RowBatch FETCHED = new SignalingRowBatch();
    public static final RowBatch FETCHED_DONE = new SignalingRowBatch();
    public static final RowBatch WAIT = new SignalingRowBatch();

    private SignalingRowBatch() {
    }

    ...

}
```

Reply:
How do you support multiple batch implementations in this case (List-based, byte-array-based, single row, etc)?

Reply:
One more thing here - all operators produce iteration results (FETCHED, WAIT, etc.), but not all operators produce batches. Specifically, top terminal operators, root and senders, never produce batches. They only signal the execution environment whether they made any progress and whether they finished.

Reply:
Ah, I see, `FETCHED` and `FETCHED_DONE` are actually providing the data, not just the signaling :)